Allow for delocating top level modules.

HexDecimal opened the pull request:
This delocates libraries into one directory per wheel.  I've added a top-level package-less wheel to test with.

Prevents copying duplicate libraries when a wheel has multiple packages.  Fixes #35.

Fixes issues where the module to work on is at the top-level, such as a wheel with no packages.  Generally any current issue where delocate mysteriously doesn't bundle libraries has been fixed.  Fixes #72, fixes #22, fixes #45, fixes #63, fixes #121, fixes #66, fixes #49, fixes #67.

See `_decide_dylib_bundle_directory` for how the folder to use is determined.  I've followed the advice from #72 to use an auditwheel-style name.  Wheels without packages use the folder `{package_name}.dylibs`, wheels with packages will put a `.dylibs` folder in only one of the packages preferring the one that matches the wheel name.  The one directory will hold the dependencies of the entire wheel.

After that the fix was simple:  In `delocate_wheel` call `delocate_path` only once and call it with a `tree_path` that includes the entire wheel.

Since this always collects everything regardless of if a package is found or not.  This will also resolve issues where namespace packages couldn't be delocated.  Fixes #95.

HexDecimal committed:
Delocate libraries into one directory per wheel.

Add a top-level package-less wheel to test.

Fixes duplicate library issues when a wheel has multiple packages.

Fixes issues where a wheel has no packages.

matthew-brett commented on delocate/delocating.py:
```diff
@@ -450,6 +450,37 @@ def _merge_lib_dict(d1, d2):
+        if directory.endswith(package_name):
+            # Prefer using the directory with the same name as the package.
+            return pjoin(directory, lib_sdir)
+    if package_dirs:
```
Just a personal preference so I don't have to think about what True might mean in the bare case:

```suggestion
    if len(package_dirs):
```

HexDecimal replied:
I never got confused by this truth test.  It's handled consistently across Python that sequences are True when they have any items and False when empty. PEP8 also recommended not adding `len` to sequences, which is likely why I got used to doing it this way.

matthew-brett replied:
Your call.  For me, I'm thinking - could this be None?  Or an integer?

HexDecimal replied:
If your IDE supports type hinting then it will plainly tell you that `package_dirs` is `Set[str]`.  If the IDE only shows the docstrings of functions then looking at `find_package_dirs` will tell you that it returns `Set[str]`.  Writing code with mystery types should become a thing of the past.

`None` and `int` are ruled out just from how the name is being used in the for loop above and in the `min` function within the if branch.  It's clearly a sequence, but without type hinting you wouldn't be able to tell if it was a sequence of `str` or something else, such as `pathlib.Path`.

matthew-brett replied:
Yes, sure, I could use other information by looking around, or if I had the right IDE with type hints (I don't), but with `len(foo)` I know immediately by looking at it...

matthew-brett commented on delocate/delocating.py:
```diff
@@ -526,42 +558,44 @@ def delocate_wheel(
-                ignore_missing=ignore_missing,
-            )
-            if copied_libs and lib_path_exists:
+        # Assume the package name from the wheel filename.
```
Function getting a bit long now.  Can you think of a way of splitting this up?  Don't worry if not - it's hardly your changes that are causing the length.

HexDecimal replied:
Everything was in one big for loop and taking out the loop caused a dedent on a lot of code.  I thought of splitting it up but in the end I was focusing more on fixing this rather major issue rather than cleaning things up.

matthew-brett replied:
Yes, no problem.  For a later date, maybe.

HexDecimal replied:
If I don't clean this up now then I'm not sure it will ever happen.

matthew-brett replied:
Totally your call.

matthew-brett commented on delocate/delocating.py:
```diff
@@ -450,6 +450,37 @@ def _merge_lib_dict(d1, d2):
+    package_name : str
+        The name of the package.
+    lib_sdir : str, optional
+        This value passed in via :func:`delocate_wheel`.
```
```suggestion
        Default value for lib sub-directory passed in via :func:`delocate_wheel`. 
        Ignored if wheel has no package directories.
```

matthew-brett commented on delocate/delocating.py:
```diff
@@ -479,6 +510,7 @@ def delocate_wheel(
     lib_sdir : str, optional
         Subdirectory name in wheel package directory (or directories) to store
         needed libraries.
+        This may be ignored depending on how the wheel is structured.
```
```suggestion
       Ignored if the wheel has no package directories, and only contains stand-alone
       modules.
```

matthew-brett commented on delocate/delocating.py:
```diff
@@ -526,42 +558,44 @@ def delocate_wheel(
+            wheel_dir, package_name, lib_sdir
+        )
+        lib_path = pjoin(wheel_dir, lib_sdir)
+        lib_path_exists = exists(lib_path)
```
Not your change, but I guess this should go above the following line, where it is first used.

HexDecimal replied:
I thought `lib_path_exists` was put before `delocate_path` to tell if the `lib_path` existed before the call was made.  Maybe a better name would work.

matthew-brett replied:
Well - let's leave for now.  But we aren't using the value for the `delocate_path` call.

matthew-brett commented on delocate/delocating.py:
```diff
@@ -526,42 +558,44 @@ def delocate_wheel(
+        )
+        if copied_libs and lib_path_exists:
+            raise DelocationError(
+                '{0} already exists in wheel but need to copy '
+                '{1}'.format(lib_path, '; '.join(copied_libs)))
```
It's time!

```suggestion
                'f{lib_path} already exists in wheel but need to copy ' +
                "; ".join(copied_libs))
```

matthew-brett commented on delocate/delocating.py:
```diff
@@ -526,42 +558,44 @@ def delocate_wheel(
+            bads = check_archs(copied_libs, require_archs, stop_fast)
+            if len(bads) != 0:
+                if check_verbose:
+                    print(bads_report(bads, pjoin(tmpdir, 'wheel')))
```
Should this be a logger call now?

HexDecimal replied:
I don't know.  It is kind of weird to optionally report error information like this when an exception is going to be raised anyway.  I might've included this in the exception message.

matthew-brett replied:
OK - how about appending this to the error message instead?

HexDecimal replied:
The current behavior is according to the documentation.  There isn't much that can be done here.

matthew-brett replied:
It's really not important, feel free to leave as is.  If, by documentation, you just meant the docstring, then I think adding it to the error message ends up with just as much "printing" as using `print`, so I don't worry about that myself.

HexDecimal replied:
Logging sends to `stderr` instead of `stdout` and changes the format.  Using print like this is technically a contract with other CLI tools which is why I'm having a hard time with it.

Since you didn't intend it to be that way then I'll be okay with breaking it.

HexDecimal replied:
I kind of want to ignore `check_verbose` and always print the full error if an issue is found.

matthew-brett replied:
Yes, that's fine too.

matthew-brett commented on delocate/tests/test_wheelies.py:
```diff
@@ -342,3 +343,22 @@ def ignore_libextfunc2(path: str) -> bool:
+
+    with InTemporaryDirectory():
+        # The module was set to expect its dependency in the libs/ directory
+        os.symlink(DATA_PATH, 'libs')
```
I think this won't work as expected on Windows, should we ever get there : https://docs.python.org/3/library/os.html

HexDecimal replied:
I knew it was Unix specific, and I've already been using symlinks for the rpath tests.

Adding support for Window's DLL's to delocate would just be writing an entire new library on top of the existing one.  I don't really see much code that would be shared between them since DLL's are more different to dylibs than Linux so's are, and those were already handled well by auditwheel.

Switching symlink out for something more cross-platform is still a reasonable idea, so I'll do that.

matthew-brett reviewed (commented):
Fantastic - thanks!    Just a few suggestions.   Are you testing the case for more than one package in wheel?

HexDecimal committed:
Apply suggestions from code review

Co-authored-by: Matthew Brett <matthew.brett@gmail.com>

HexDecimal committed:
Copy rpath libraries in tests instead of symlinking them.

matthew-brett commented:
How about a test for namespace packages as well?  (Sorry for the stream of requests).

HexDecimal committed:
Refactor decloate_wheel function.

Move _make_install_name_ids_unique outside of the function.

Remove all_copied since this function no longer has a loop.

HexDecimal commented:
> Are you testing the case for more than one package in wheel?
> How about a test for namespace packages as well?

Only the current wheels are tested so more wheels either need to be added or the current need to be modified to be more diverse.  There are no wheels using namespace packages, or wheels with more than one package.

Namespace packages are currently treated the same as wheels with no packages.  I see this as acceptable since wheel names shouldn't collide.

While there are no wheels with multiple packages, both the conditions of selecting a package by the wheel name and selecting a package alphabetically have been covered by tests.

HexDecimal committed:
Clean up check architecture code.

HexDecimal committed:
Enable Pytest doctests.

HexDecimal committed:
Rewrite wheel architecture tests.

HexDecimal committed:
Remove unused _merge_lib_dict function.

HexDecimal committed:
Add namespace wheel test.

HexDecimal commented:
I've added a namespace wheel with a typical test for it.

matthew-brett commented on delocate/delocating.py:
```diff
@@ -437,17 +437,72 @@ def _copy_lib_dict(lib_dict):
+def _make_install_name_ids_unique(
+    libraries: Iterable[str], install_id_prefix: str
+) -> None:
+    """Replace each libraries install name ids with a unique id.
```
```suggestion
    """Replace each library's install name id with a unique id.
```

matthew-brett commented on delocate/delocating.py:
```diff
@@ -499,7 +556,7 @@ def delocate_wheel(
         (e.g "i386" or "x86_64").
     check_verbose : bool, optional
-        If True, print warning messages about missing required architectures
+        This flag is currently ignored.
```
I guess we should decide on deprecation here.  Maybe we could set this to None by default, and raise a deprecation warning if it is not None.    Maybe warn too that this and subsequent arguments will become keyword-only in the future.

HexDecimal replied:
It can be hard to switch a function over to keyword only arguments.  Only `FutureWarning` is loud enough that you can expect people using the older API to notice that it's actually deprecated.

HexDecimal replied:
I've marked it with a DeprecationWarning and labeled the remaining parameters as keyword-only.

matthew-brett replied:
Thanks.  I'll follow up with a little PR to check the further parameters.

matthew-brett reviewed (commented):
Just a final few comments.

HexDecimal committed:
Fix spelling of _make_install_name_ids_unique docstring.

Co-authored-by: Matthew Brett <matthew.brett@gmail.com>

HexDecimal committed:
Mark check_verbose parameter as deprecated.

This parameter is no longer used by its function and may be removed in
the future.

HexDecimal commented:
I'd like to make a release of a version with this merged.  [make_release.rst](https://github.com/matthew-brett/delocate/blob/dbb9367bcf2b8519199587b5336bcbd2d12e3c82/doc/devel/make_release.rst) mentions making announcements to a mailing list, but I don't see the mailing list this is referring to.  If you ever want to make a release then all you have to do right now is push a tag with the version.

We should also look at the other PR's which were trying to implement this feature: #36 #38 #39 #74 can probably be closed without merging.  #96 might improve the handling of namespace packages but I don't think the author will care much now that namespace packages decloate correctly.

HexDecimal commented:
I'm not sure.  It's probably not as big of a deal to have a stable programming API for a primarily CLI tool, but my current suggestion is to go with `0.10.0` as the next version.

matthew-brett commented:
That's fine too - it's really not a big deal - I don't think people pay much attention to version numbers anyway.

HexDecimal commented:
Semantic versioning just sets the rules for why versions are what they are.  If you released `1.0.0` and then break the API right away then the next version will be `2.0.0` and it isn't really a big deal for that to happen as long as the versions are set correctly.

`1.0.0` is used to express confidence that the current API is stable and that more care will be taken to keep compatibility with it. I wouldn't use it to mean anything else.  With how careful you've been treating breaking cases you probably could've been using `1.0.0` a long time ago.

Personally, there are things I'd like to do or braking changes I'd consider making before `1.0.0`.  I'd rather not release that version before everything is type-hinted for example.

asottile commented:
can confirm this works -- thanks for picking up this patch!
